src/Compilers/CSharp/Portable/Binder/Semantics/AccessCheck.cs

```diff
@@ -511,7 +530,75 @@ private static bool IsNamedTypeAccessible(NamedTypeSymbol type, Symbol within, r
             }
 
             visited?.Free();
+
+            if (!result && baseTypeIsInterface)
+            {
+                Debug.Assert(!result);
+
+                while (baseInterfaces.Count != 0)
```

Comment:
>while [](start = 16, length = 5)

This doubly-nested loop appears to be more complex than necessary, and repeating work that can be accessed by a single call to
``` c#
        internal ImmutableArray<NamedTypeSymbol> TypeSymbol.AllInterfacesWithDefinitionUseSiteDiagnostics(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```
If you do that, you don't need to deduplicate different instantiations of the same base interface: it is rare, but you still get the correct result checking the type again.  However, in the common case where there is no duplication, this code does unnecessary filtering using a hash set.

Reply:
> If you do that, you don't need to deduplicate different instantiations of the same base interface: it is rare, but you still get the correct result checking the type again.  However, in the common case where there is no duplication, this code does unnecessary filtering using a hash set.

```AllInterfacesWithDefinitionUseSiteDiagnostics``` causes cycles. 


---
In reply to: [264416410](https://github.com/dotnet/roslyn/pull/34012#discussion_r264416410) [](ancestors = 264416410)

Reply:
I'm not sure how it could (assuming you're calling it precisely once here) as it has cycle checks.

---
In reply to: [264418856](https://github.com/dotnet/roslyn/pull/34012#discussion_r264418856) [](ancestors = 264418856,264416410)

Reply:
> I'm not sure how it could (assuming you're calling it precisely once here) as it has cycle checks.

It doesn't pay attention to ```basesBeingResolved``` and we can and do get here again once we call ```AllInterfacesWithDefinitionUseSiteDiagnostics ``` from here. For the same exact reason we do not use BaseTypeWithDefinitionUseSiteDiagnostics in this function to get a base type.

---
In reply to: [264423139](https://github.com/dotnet/roslyn/pull/34012#discussion_r264423139) [](ancestors = 264423139,264418856,264416410)

Reply:
Is there a test that would trigger that infinite recursion?

---
In reply to: [264425213](https://github.com/dotnet/roslyn/pull/34012#discussion_r264425213) [](ancestors = 264425213,264423139,264418856,264416410)

Reply:
> Is there a test that would trigger that infinite recursion?

Yes.

---
In reply to: [264431511](https://github.com/dotnet/roslyn/pull/34012#discussion_r264431511) [](ancestors = 264431511,264425213,264423139,264418856,264416410)

Reply:
Can you please identify that test by telling me its name?

---
In reply to: [264439678](https://github.com/dotnet/roslyn/pull/34012#discussion_r264439678) [](ancestors = 264439678,264431511,264425213,264423139,264418856,264416410)

Reply:
> Can you please identify that test by telling me its name?

I think it is at least ```NestedTypes_04``` and probably other ```NestedTypes_*``` tests added/modified in this PR.

---
In reply to: [264906559](https://github.com/dotnet/roslyn/pull/34012#discussion_r264906559) [](ancestors = 264906559,264439678,264431511,264425213,264423139,264418856,264416410)